### ISSUE TITLE
memory leak in `podman build --layers` when using ADD from network

### ISSUE THREAD

**Issue by dustymabe** [containers/libpod#1276]
/kind bug

**Description**

Trying to build with --layers seems to cause a memory leak when using ADD from the network (i.e. `ADD http://`). If I use add from local then I don't have the problem. Here is the Dockerfile:

```
FROM registry.fedoraproject.org/fedora:27

LABEL RUN sudo podman run -it -v $PWD/logs/:/home/user/.weechat/logs/:z --name weechat weechat

RUN dnf install -y weechat python2-websocket-client

ENV HOME /home/user
RUN useradd -m user

RUN mkdir -p /home/user/.weechat/python/autoload/
ADD https://raw.githubusercontent.com/wee-slack/wee-slack/v2.0.0/wee_slack.py /home/user/.weechat/python/autoload/wee_slack.py
ADD start-script.txt /home/user/start-script.txt

RUN chown -R user $HOME

VOLUME /home/user/.weechat/logs/

WORKDIR $HOME
USER user

ENV TZ='America/New_York'

CMD weechat -r "$(cat start-script.txt  | grep -v '^#' | tr '\n' ';')"
```

Here is the output:

```
[dustymabe@dhcp137-98 weechat]$ bash build.sh                                                                                                                                                                                                                                 
+ sudo podman build -t weechat --layers ./                                                                                                                                                                                                                                    
STEP 1: FROM registry.fedoraproject.org/fedora:27                                                                                                                                                                                                                             
STEP 2: LABEL RUN sudo podman run -it -v $PWD/logs/:/home/user/.weechat/logs/:z --name weechat weechat
--> Using cache b70cd21686b8ae71cfd91b9bef87189da60dc72ada7e814635899ec3c05b4508                                                                                                                                                                                              
STEP 3: FROM b70cd21686b8ae71cfd91b9bef87189da60dc72ada7e814635899ec3c05b4508
STEP 4: RUN dnf install -y weechat python2-websocket-client                                                                                                                                                                                                                   
--> Using cache 8a5c77ee758d52121e18b82721aeae2c41c005288b337121348aa7c93b839fec                                                                                                                                                                                              
STEP 5: FROM 8a5c77ee758d52121e18b82721aeae2c41c005288b337121348aa7c93b839fec                                                                                                                                                                                                 
STEP 6: ENV HOME /home/user
--> Using cache 7e55665d81548f3a3f7b1e83728875c40914d87aac71020d49cd7c8e13c4cdff                                                                                                                                                                                              
STEP 7: FROM 7e55665d81548f3a3f7b1e83728875c40914d87aac71020d49cd7c8e13c4cdff
STEP 8: RUN useradd -m user                                                                                                                                                                                                                                                   
--> Using cache 78e89927417e4a8725a61afcfb64474f2a0d9ab5e10496c6a29eb67a5f80b2ea
STEP 9: FROM 78e89927417e4a8725a61afcfb64474f2a0d9ab5e10496c6a29eb67a5f80b2ea
STEP 10: RUN mkdir -p /home/user/.weechat/python/autoload/
--> Using cache ed0d892aee6f78cab156d4d941179a8364d6cb8478daffff27592b1985078ad4
STEP 11: FROM ed0d892aee6f78cab156d4d941179a8364d6cb8478daffff27592b1985078ad4
STEP 12: ADD https://raw.githubusercontent.com/wee-slack/wee-slack/v2.0.0/wee_slack.py /home/user/.weechat/python/autoload/wee_slack.py
fatal error: runtime: out of memory
                                                                                                                                                                                                                                                                              
runtime stack:
runtime.throw(0x12357f9, 0x16)                                                                                                                                                                                                                                                
        /usr/lib/golang/src/runtime/panic.go:616 +0x81
runtime.sysMap(0xc7d1030000, 0x15eb30000, 0x0, 0x1d6deb8)
        /usr/lib/golang/src/runtime/mem_linux.go:216 +0x20a
runtime.(*mheap).sysAlloc(0x1d53cc0, 0x15eb30000, 0x8)
        /usr/lib/golang/src/runtime/malloc.go:470 +0xd4
runtime.(*mheap).grow(0x1d53cc0, 0xaf598, 0x0)
        /usr/lib/golang/src/runtime/mheap.go:907 +0x60
runtime.(*mheap).allocSpanLocked(0x1d53cc0, 0xaf598, 0x1d6dec8, 0x1)
        /usr/lib/golang/src/runtime/mheap.go:820 +0x301
runtime.(*mheap).alloc_m(0x1d53cc0, 0xaf598, 0xffffffffffff0100, 0x7f3949ffabb8)
        /usr/lib/golang/src/runtime/mheap.go:686 +0x118
runtime.(*mheap).alloc.func1()
        /usr/lib/golang/src/runtime/mheap.go:753 +0x4d                                                                                                                                                                                                                        
runtime.(*mheap).alloc(0x1d53cc0, 0xaf598, 0x7f3949010100, 0x43ce67)                                                                                                                                                                                                          
        /usr/lib/golang/src/runtime/mheap.go:752 +0x8a
runtime.largeAlloc(0x15eb30000, 0x7f3949ff0001, 0x43d036)                                                                                                                                                                                                                     
        /usr/lib/golang/src/runtime/malloc.go:826 +0x94
runtime.mallocgc.func1()                                                                                                                                                                                                                                                      
        /usr/lib/golang/src/runtime/malloc.go:721 +0x46
runtime.systemstack(0xc400000000)                                                                                                                                                                                                                                             
        /usr/lib/golang/src/runtime/asm_amd64.s:409 +0x79
runtime.mstart()
        /usr/lib/golang/src/runtime/proc.go:1175

goroutine 1 [running]:                                                                                                                                                                                                                                                         
runtime.systemstack_switch()                                                                                                                                                                                                                                                   
        /usr/lib/golang/src/runtime/asm_amd64.s:363 fp=0xc420553d08 sp=0xc420553d00 pc=0x45f9a0
runtime.mallocgc(0x15eb30000, 0x1053960, 0xc420553e01, 0xc4200c3b04)
        /usr/lib/golang/src/runtime/malloc.go:720 +0x8a2 fp=0xc420553da8 sp=0xc420553d08 pc=0x41bcc2
runtime.growslice(0x1053960, 0xc64406a000, 0x1188f200, 0x1188f200, 0x1188f201, 0xc64406a000, 0xe072800, 0x1188f200)
        /usr/lib/golang/src/runtime/slice.go:179 +0x14a fp=0xc420553e10 sp=0xc420553da8 pc=0x44bb7a
github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah.(*Executor).getFilesToCopy(0xc4203efb80, 0xc420255dc0, 0x0, 0x136b4c0, 0xc420454cc0, 0xc420553f48, 0x4e46a9)                                                                             
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah/build.go:950 +0xfd fp=0xc420553ee8 sp=0xc420553e10 pc=0xe65add                                  
github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah.(*Executor).copiedFilesMatch(0xc4203efb80, 0xc420255dc0, 0xc420454fa0, 0xc4201ee240, 0x7, 0x9)                                                                                           
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah/build.go:972 +0x91 fp=0xc420553fa0 sp=0xc420553ee8 pc=0xe65f21                                  
github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah.(*Executor).layerExists(0xc4203efb80, 0x135c4e0, 0xc4200c4048, 0xc420255dc0, 0xc420108300, 0x5, 0x10, 0x0, 0x0, 0x0, ...)                                                                
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah/build.go:882 +0x2f3 fp=0xc4205541c0 sp=0xc420553fa0 pc=0xe650f3                                 
github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah.(*Executor).Execute(0xc4203efb80, 0x135c4e0, 0xc4200c4048, 0xc4205282c0, 0xc420526620, 0xc42043d055, 0x24)                                                                               
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah/build.go:772 +0xa58 fp=0xc4205543d0 sp=0xc4205541c0 pc=0xe644b8                                 
github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah.(*Executor).Build(0xc4203ef900, 0x135c4e0, 0xc4200c4048, 0xc4205200c0, 0x1, 0x1, 0x0, 0x0)                                                                                               
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah/build.go:1137 +0x1f5 fp=0xc420554498 sp=0xc4205543d0 pc=0xe689d5                                
github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah.BuildDockerfiles(0x135c4e0, 0xc4200c4048, 0x13773a0, 0xc4203b3d40, 0xc4200e18e0, 0x1b, 0x0, 0x0, 0x0, 0x0, ...)                                                                          
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/projectatomic/buildah/imagebuildah/build.go:1226 +0x1009 fp=0xc4205548d8 sp=0xc420554498 pc=0xe69be9                               
github.com/projectatomic/libpod/libpod.(*Runtime).Build(0xc4200fadc0, 0x135c4e0, 0xc4200c4048, 0xc4200e18e0, 0x1b, 0x0, 0x0, 0x0, 0x0, 0x0, ...)                                                                                                                              
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/libpod/runtime_img.go:193 +0xc7 fp=0xc420554b48 sp=0xc4205548d8 pc=0xeb97b7                                                                          
main.buildCmd(0xc420144420, 0x0, 0x0)                                                                 
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/cmd/podman/build.go:220 +0x10f6 fp=0xc4205552c8 sp=0xc420554b48 pc=0xf38756                                                                          
github.com/projectatomic/libpod/vendor/github.com/urfave/cli.HandleAction(0x1071260, 0x12b0610, 0xc420144420, 0x0, 0xc420030480)
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/urfave/cli/app.go:501 +0xc8 fp=0xc4205552f0 sp=0xc4205552c8 pc=0xed8888                                                            
github.com/projectatomic/libpod/vendor/github.com/urfave/cli.Command.Run(0x122241f, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x12649cb, 0x32, 0x0, ...)                                                                                                                                  
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/urfave/cli/command.go:165 +0x47d fp=0xc420555498 sp=0xc4205552f0 pc=0xed952d                                                       
github.com/projectatomic/libpod/vendor/github.com/urfave/cli.(*App).Run(0xc4201aaa80, 0xc4200cc000, 0x6, 0x6, 0x0, 0x0)
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/urfave/cli/app.go:259 +0x6e8 fp=0xc420555680 sp=0xc420555498 pc=0xed6a88                                                           
main.main()                                                                  
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/cmd/podman/main.go:214 +0x1758 fp=0xc420555f88 sp=0xc420555680 pc=0xf55448                                                                           
runtime.main()                                                                  
        /usr/lib/golang/src/runtime/proc.go:198 +0x212 fp=0xc420555fe0 sp=0xc420555f88 pc=0x436862
runtime.goexit()                                          
        /usr/lib/golang/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc420555fe8 sp=0xc420555fe0 pc=0x462531
                                                                              
goroutine 19 [chan receive]:                                                                                                           
github.com/projectatomic/libpod/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x1d4c860)
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/golang/glog/glog.go:882 +0x8b                                                                                                      
created by github.com/projectatomic/libpod/vendor/github.com/golang/glog.init.0
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/golang/glog/glog.go:410 +0x203                                                                                                     
                                                      
goroutine 20 [syscall]:                                  
os/signal.signal_recv(0x0)                                 
        /usr/lib/golang/src/runtime/sigqueue.go:139 +0xa6
os/signal.loop()                                       
        /usr/lib/golang/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0                           
        /usr/lib/golang/src/os/signal/signal_unix.go:28 +0x41       
goroutine 6 [syscall]:                                
syscall.Syscall6(0xe8, 0xa, 0xc42032dc3c, 0x7, 0xffffffffffffffff, 0x0, 0x0, 0x2, 0x2, 0x0)
        /usr/lib/golang/src/syscall/asm_linux_amd64.s:44 +0x5
github.com/projectatomic/libpod/vendor/golang.org/x/sys/unix.EpollWait(0xa, 0xc42032dc3c, 0x7, 0x7, 0xffffffffffffffff, 0x433ff0, 0xc42032dc38, 0x1)
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/golang.org/x/sys/unix/zsyscall_linux_amd64.go:1520 +0x7a
github.com/projectatomic/libpod/vendor/github.com/fsnotify/fsnotify.(*fdPoller).wait(0xc420314960, 0x12af500, 0xc420314960, 0x0)
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/fsnotify/fsnotify/inotify_poller.go:86 +0x91
github.com/projectatomic/libpod/vendor/github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc42040eb40)
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/fsnotify/fsnotify/inotify.go:192 +0x194
created by github.com/projectatomic/libpod/vendor/github.com/fsnotify/fsnotify.NewWatcher
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/fsnotify/fsnotify/inotify.go:61 +0x203
                                                             
goroutine 7 [select]:                                                                                                                                                                                                                                                         
github.com/projectatomic/libpod/vendor/github.com/cri-o/ocicni/pkg/ocicni.(*cniNetworkPlugin).monitorConfDir(0xc4200fafd0, 0xc42043ae80)                                                                                                                                      
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/cri-o/ocicni/pkg/ocicni/ocicni.go:142 +0x1b7
created by github.com/projectatomic/libpod/vendor/github.com/cri-o/ocicni/pkg/ocicni.initCNI                                                                                                                                                                                  
        /builddir/build/BUILD/libpod-80612fb046f38d1a72fe49c6c3c01fa5e0bca54a/_build/src/github.com/projectatomic/libpod/vendor/github.com/cri-o/ocicni/pkg/ocicni/ocicni.go:234 +0x38b
```

During the build I see podman running away with memory:


```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                                                                                                    
 1838 root      20   0 12.019g 8.864g  24032 S 188.0 56.9   1:07.56 podman
```


** more details **

```
[dustymabe@dhcp137-98 weechat]$ sudo podman version
[sudo] password for dustymabe:
Version:       0.7.4
Go Version:    go1.10.3
OS/Arch:       linux/amd64
[dustymabe@dhcp137-98 weechat]$
[dustymabe@dhcp137-98 weechat]$ sudo podman version
Version:       0.7.4
Go Version:    go1.10.3
OS/Arch:       linux/amd64
[dustymabe@dhcp137-98 weechat]$ sudo podman info
host:
  MemFree: 15084412928
  MemTotal: 16716890112
  SwapFree: 2943746048
  SwapTotal: 4294963200
  arch: amd64
  cpus: 8
  hostname: dhcp137-98.rdu.redhat.com
  kernel: 4.17.11-200.fc28.x86_64
  os: linux
  uptime: 25m 23.76s
insecure registries:
  registries: []
registries:
  registries:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ContainerStore:
    number: 3
  GraphDriverName: overlay
  GraphOptions:
  - overlay.override_kernel_check=true
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
  ImageStore:
    number: 27
  RunRoot: /var/run/containers/storage

[dustymabe@dhcp137-98 weechat]$ rpm-ostree status
State: idle; auto updates enabled (check; no runs since boot)
Deployments:
● ostree://unifiedrepo:fedora/28/x86_64/atomic-host
                   Version: 28.20180804.0 (2018-08-04 19:52:51)
                BaseCommit: 5633f5b369b166d104720a4da31da1ac6b5e3c0de5c5bae70cb40a850dede502
              GPGSignature: Valid signature by 128CF232A9371991C8A65695E08E7E629DB62FB1
           LayeredPackages: aria2 git git-annex mosh pciutils tig vim

  ostree://unifiedrepo:fedora/28/x86_64/atomic-host
                   Version: 28.20180722.0 (2018-07-23 00:38:05)
                BaseCommit: 106a7095fd46741948ba60dbd0586668ae7abe6336671444ec790d7541a88778
              GPGSignature: Valid signature by 128CF232A9371991C8A65695E08E7E629DB62FB1
           LayeredPackages: aria2 git git-annex mosh pciutils tig vim

  ostree://fedora-atomic-27:fedora/27/x86_64/atomic-host
                   Version: 27.122 (2018-04-18 23:34:24)
                BaseCommit: 931ebb3941fc49af706ac5a90ad3b5a493be4ae35e85721dabbfd966b1ecbf99
              GPGSignature: Valid signature by 860E19B0AFA800A1751881A6F55E7430F5282EE4
           LayeredPackages: aria2 git git-annex mosh pciutils tig vim weechat

```

**Comment by TomSweeneyRedHat** [comment]
@dustymabe thanks for the problem report.  Just to make sure I'm on the same page, if you change this line in the Dockerfile: 

```
ADD https://raw.githubusercontent.com/wee-slack/wee-slack/v2.0.0/wee_slack.py /home/user/.weechat/python/autoload/wee_slack.py
```
to
```
ADD ./wee_slack.py /home/user/.weechat/python/autoload/wee_slack.py
```
it works? Did you by chance try doing the bud command without the `--layers` param?  Did that work?

**Comment by dustymabe** [comment]
> it works?

correct

>  Did you by chance try doing the bud command without the --layers param? Did that work?

just tried without `--layers`. I do not see this bug without `--layers`.

**Comment by TomSweeneyRedHat** [comment]
Thanks for the follow up @dustymabe, that will help narrow the search down.

**Comment by TomSweeneyRedHat** [comment]
@umohnani8 any thoughts?

**Comment by umohnani8** [comment]
@TomSweeneyRedHat will look into it.

**Comment by rhatdan** [comment]
@dustymabe I just tried this out with the latest podman and it worked for me.  Could you attempt this again? with podman 0.8.3 or later?

**Comment by dustymabe** [comment]
just tried with `podman-0.8.3-1.git9d09a4d.fc28.x86_64` rpm and I got a segfault

```
[vagrant@vanilla-f28 weechat]$ bash build.sh 
+ sudo podman build -t weechat --layers ./
STEP 1: FROM registry.fedoraproject.org/fedora:28
Getting image source signatures
Copying blob sha256:6b8551f02cfd71ef7b00fc994a1c573c53409552504fc0edd4cc14164e2e40d9
 85.23 MB / 85.23 MB [======================================================] 5s
Copying config sha256:e555121ced0fcad9197d7d0445daff0e42d8f0e0c37362b66b817b8713dcbb3a
 1.27 KB / 1.27 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
STEP 2: LABEL RUN sudo podman run -it -v $PWD/logs/:/home/user/.weechat/logs/:z --name weechat weechat
--> ffd8bfcff2f7daf20c49ed57f69bd9ab1358a6e70610b9f5cdf8d5d525f604de
STEP 3: FROM ffd8bfcff2f7daf20c49ed57f69bd9ab1358a6e70610b9f5cdf8d5d525f604de
STEP 4: RUN dnf install -y weechat python2-websocket-client
container exited on segmentation fault
error building at step {Env:[DISTTAG=f28container FGC=f28 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[dnf install -y weechat python2-websocket-client] Flags:[] Attrs:map[] Message:RUN dnf install -y weechat python2-websocket-client Original:RUN dnf install -y weechat python2-websocket-client}: exit status 1
```

Note I ran `sudo restorecon -R -v /var/lib/containers` to make sure I wasn't hitting any issues like https://github.com/containers/libpod/issues/1118#issuecomment-411388062

**Comment by umohnani8** [comment]
@rhatdan @dustymabe just tried it with buildah 1.4-dev and podman 0.8.2.1 and the build works fine for me also. Tried it with the exact Dockerfile provided in this issue.

**Comment by umohnani8** [comment]
@dustymabe thanks for the info. I found the issue, opening a PR with the fix.

**Comment by umohnani8** [comment]
Being fixed in https://github.com/projectatomic/buildah/pull/947, will vendor into podman once it is merged into buildah.

**Comment by umohnani8** [comment]
Fixed in #1330
